filclient/filclient.go

```diff
@@ -853,221 +860,137 @@ func (fc *FilClient) RetrieveContent(ctx context.Context, miner address.Address,
 		return nil, err
 	}
 
-	pchaddr, err := fc.getPaychWithMinFunds(ctx, minerOwner)
+	// Get the payment channel and create a lane for this retrieval
+	pchAddr, err := fc.getPaychWithMinFunds(ctx, minerOwner)
 	if err != nil {
 		return nil, xerrors.Errorf("failed to get payment channel: %w", err)
 	}
-
-	// Allocate a lane on our payment channel, usually you want a new lane per retrieval
-	lane, err := fc.pchmgr.AllocateLane(pchaddr)
+	pchLane, err := fc.pchmgr.AllocateLane(pchAddr)
 	if err != nil {
 		return nil, xerrors.Errorf("failed to allocate lane: %w", err)
 	}
 
-	// Use the data transfer protocol to propose the retrieval deal
-	sel := shared.AllSelector()
-	var vouch datatransfer.Voucher = proposal
-	chanid, err := fc.dataTransfer.OpenPullDataChannel(ctx, mpid, vouch, proposal.PayloadCID, sel)
-	if err != nil {
-		return nil, err
-	}
+	var chanid datatransfer.ChannelID
+	var chanidMu sync.Mutex
 
-	unsub := fc.dataTransfer.SubscribeToEvents(func(event datatransfer.Event, channelState datatransfer.ChannelState) {
-		if channelState.ChannelID() == chanid {
-			fmt.Printf("[%s] event: %s %d %s %s\n", time.Now().Format("15:04:05"), event.Message, event.Code, datatransfer.Events[event.Code], event.Timestamp.Format("15:04:05"))
-			fmt.Printf("channelstate: %s %s\n", datatransfer.Statuses[channelState.Status()], channelState.Message())
-		}
-	})
-	defer unsub()
+	// Set up incoming events handler
 
-	// NB: data transfer will propose the retrieval, and if the miner accepts
-	// it, will start sending data via graphsync to us. This happens behind the
-	// scenes, and we dont get much introspection into this
+	// The next nonce (incrementing unique ID starting from 0) for the next voucher
+	var nonce uint64 = 0
 
-	// Now, we poll the data transfer channel status and wait until it says its accepted
-	if err := fc.waitForDealAccepted(ctx, chanid); err != nil {
-		return nil, err
-	}
+	// Used to limit prints to the console about received status
+	lastReceivedUpdate := time.Now()
 
-	var nonce uint64
-	total := abi.NewTokenAmount(0)
+	// dtRes receives either an error (failure) or nil (success) which is waited
+	// on and handled below before exiting the function
+	dtRes := make(chan error)
 
-	var voucherCount int
-	for {
-		// Now that the deal has been accepted, we sit around and wait for data
-		// to come in and for the miner to ask for more money
-		amt, nvc, done, err := fc.waitForPaymentNeeded(ctx, chanid, voucherCount)
-		if err != nil {
-			return nil, xerrors.Errorf("waitForPayment needed failed: %w", err)
-		}
+	unsubscribe := fc.dataTransfer.SubscribeToEvents(func(event datatransfer.Event, state datatransfer.ChannelState) {
+		// Lock chanid for the entire callback
+		chanidMu.Lock()
```

Comment:
instead of locking around the entire function, we could just lock around checking if this is the right channel. Something like:
```go
chanidMu.Lock()
match := state.ChannelID() == chanid
chanidMu.Unlock()

if !match {
  return
}
...
```

Reply:
chanid actually gets used later in the function, so i can't keep the locked section this short. i alternatively considered copying chanid as chanidLocked or something along those lines and then replacing uses of chanid later in the function with that, but i realized that it's not really useful because the only time outside of this function that the mutex gets locked is to set chanid initially. 

should i go back to that way?

Reply:
Ah, good point about using it later down. Taking a copy under the lock should work well enough then. Its an annoying but technically correct approach.